docs/architecture/adr-007.md

```diff
@@ -0,0 +1,93 @@
+# ADR 007: Generic event formats for evm-hook actions
+
+## Changelog
+* 9-06-2021: Initial Draft
+
+## Context
+
+The current evm-hooks relies on event emitted from a smart contract to trigger specific actions upon contract execution.
+The current way to define those events is not flexible: events are hardcoded, parameters fixed and correspond only to one specific action.
+
+This can lead to potential issue in the long run when commands needs to be updated to add more parameters (ibc-go/ gravity bridge) or even if we need to implement more "actions". Since contracts are immutable, contract developers won't be able to have access to these functionalities after deploying their CRC20 contracts if we perform an upgrade.
+
+In this ADR, we define generic events format that can be use by smart contract developer to call any "action" upon successful contract execution. 
+
+## Decision
+
+We define 2 specifics events that smart contract can use to trigger specific action after contract execution.
+
+```
+# Trigger an action after contract execution which only involve the sender account
+__CronosCall(string functionName, address sender,  byte[]args)
+
+# Trigger an action after contract execution that require moving asset from the smart contract "account"
+__CronosCallAfterPayment(string functionName, address payer, uint amount, byte[]args)
+
+```
+
+- **functionName** corresponds to the equivalent hardcoded eventName such as `CronosSendToIBC` for the event `__CronosSendToIbc` or `CronosSendToChain` for the event `__CronosSendToChain`
+- **payer** is the address which make the payment, usually equal to `msg.sender`
+- **amount** corresponds to the `amount` that the payer has paid (either by burning or sending asset to the SC)
+- **args** corresponds to the argument in order of the event removing the `sender` and `amount` from the parameters encoded using the abi event pack
+
+
+
+An example of smart contract implementing those function
+
+```
+pragma solidity ^0.6.8;
+
+import "ds-token/token.sol";
+
+Contract Demo is DSToken {
+
+ __CronosCallAfterPayment(string functionName, address sender, uint amount, byte[]args)
+ 
+ // unsafe_burn burn tokens without user's approval and authentication, used internally
+ function unsafe_burn(address addr, uint amount) internal {
+     // Deduct user's balance without approval
+     require(balanceOf[addr] >= amount, "ds-token-insufficient-balance");
+     balanceOf[addr] = sub(balanceOf[addr], amount);
+     totalSupply = sub(totalSupply, amount);
+     emit Burn(addr, amount);
+ }
+ 
+ // send an "amount" of the contract token to recipient through IBC
+ function send_to_ibc(byte[] recipient) public {
+     unsafe_burn(msg.sender, amount);
+     emit __CronosCallAfterPayment("CronosSendToIbc", msg.sender, amount, recipient);
```

Comment:
This example is no difference then before, what this proposal really enables is the smart contract can provide a generic API like `call_cronos` where end user can supply the `functionName` and generic args, right?

Reply:
I changed the field name to be more generic

Reply:
I think what we really do here is to solve the upgradeability issue, which is to avoid upgrading dapp's contract when the chain added new capabilities to these apis, right?
So I suggest putting more emphasis on the upgradeability issue, and demonstratee in the example.